src/librustc_typeck/check/wfcheck.rs

```diff
@@ -415,19 +404,141 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                 }
             }
 
-            let predicates = fcx.tcx.predicates_of(item_def_id).instantiate_identity(fcx.tcx);
-            let predicates = fcx.normalize_associated_types_in(item.span, &predicates);
-            this.check_where_clauses(fcx, item.span, &predicates);
+            this.check_where_clauses(fcx, item.span, item_def_id);
 
             fcx.impl_implied_bounds(item_def_id, item.span)
         });
     }
 
+    /// Checks where clauses and inline bounds.
     fn check_where_clauses<'fcx, 'tcx>(&mut self,
                                        fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
                                        span: Span,
-                                       predicates: &ty::InstantiatedPredicates<'tcx>)
+                                       def_id: DefId)
     {
+        use ty::subst::Subst;
+        use ty::Predicate;
+
+        let generics = self.tcx.generics_of(def_id);
+        let defaults = generics.types.iter().filter_map(|p| match p.has_default {
+                                                                true => Some(p.def_id),
+                                                                false => None,
+                                                        });
+        // Defaults must be well-formed.
+        for d in defaults {
+            fcx.register_wf_obligation(fcx.tcx.type_of(d), fcx.tcx.def_span(d), self.code.clone());
+        }
```

Comment:
This part seems unambiguously good.

Reply:
Though we likely want to do a crater run and decide if we need forwards compat warnings.